Python/flowgraph.c

```diff
@@ -1017,7 +1017,17 @@ remove_redundant_nops(basicblock *bb) {
                 }
                 /* or if last instruction in BB and next BB has same line number */
                 if (next) {
-                    if (lineno == next->b_instr[0].i_loc.lineno) {
+                    location next_loc = NO_LOCATION;
+                    for (int next_i=0; next_i < next->b_iused; next_i++) {
```

Comment:
Initially, I thought it would be simpler to have the same behavior whether the next instruction is in the same basicblock or the next one; i.e. it would be clearer to just match the behavior above on lines 1004-1011. If the next instruction has no location, set it to this NOP's location and remove this NOP.

But then I realized that this is not actually equivalent in the case where we are crossing to the next basicblock, because the next basicblock may be a jump target, and that jump should not report the location from the current NOP. So we do need to handle the two cases differently.

I'm not sure if this subtlety will be obvious to future readers/modifiers of this code?

We could safely use this new behavior (skip NOP without lineno that will be removed) in all cases, and restructure the code so that first we find the next "relevant" instruction (whether in current or next basic block), and then compare its location with the current NOP. I think this would be somewhat clearer, but it's up to you whether you agree / think it's worth it.

Reply:
> We could safely use this new behavior (skip NOP without lineno that will be removed) in all cases

I think we are a bit more aggressive when it's within the same basic block: we will hand the location on to any instruction that doesn't have a line number, not just a NOP.